stl/inc/flat_map

```diff
@@ -0,0 +1,869 @@
+        return const_iterator{_STD move(_Key_it), _STD move(_Val_it)};
+    }
+
+private:
+    key_compare _Key_compare;
+    containers _Data;
+
+    template <class _K1, class _K2>
+        requires (same_as<remove_cvref_t<_K1 &&>, key_type> && same_as<remove_cvref_t<_K2 &&>, key_type>)
+              || (constructible_from<key_type, _K1 &&> && constructible_from<key_type, _K2 &&>
+                  && _Compare_is_transparent<key_compare>)
+    bool _Key_equal(_K1&& _X, _K2&& _Y) const {
+        return !_Key_compare(_STD forward<_K1>(_X), _STD forward<_K2>(_Y))
+            && !_Key_compare(_STD forward<_K2>(_Y), _STD forward<_K1>(_X));
+    }
+
+    void _SortAndDedup() {
+        _Sort();
+        _Dedup();
+    }
+
+    void _Sort() {
+        _Clear_flat_map_scope_guard _Guard{this};
+        auto _Zip_view = _RANGES views::zip(_Data.keys, _Data.values);
+        _RANGES sort(_Zip_view, value_compare(_Key_compare));
+        _Guard._Clearable = nullptr;
+    }
+
+    void _Dedup() {
+        _Clear_flat_map_scope_guard _Guard{this};
+        auto _Zip_view = _RANGES views::zip(_Data.keys, _Data.values);
+        auto _Subrange = _RANGES unique(
+            _Zip_view, [this](const_reference _X, const_reference _Y) { return this->_Key_equal(_X.first, _Y.first); });
+        auto _Remaining_count = _STD distance(_Zip_view.begin(), _Subrange.begin());
+        _Data.keys.erase(_Data.keys.begin() + _Remaining_count, _Data.keys.end());
+        _Data.values.erase(_Data.values.begin() + _Remaining_count, _Data.values.end());
+        _Guard._Clearable = nullptr;
+    }
+
+    void _Insert_exact(const_iterator _Position, value_type&& _Val) {
+        _Clear_flat_map_scope_guard _Guard{this};
+        _Data.keys.insert(_Position._Key_it, _STD move(_Val.first));
+        _Data.values.insert(_Position._Mapped_it, _STD move(_Val.second));
+        _Guard._Clearable = nullptr;
+    }
+
+    template <bool _Override_if_exists, class _K, class... _Mapped_args_t>
+    iterator _Emplace_hint(const_iterator _Position, _K&& _Key, _Mapped_args_t&&... _Args)
+        requires is_constructible_v<mapped_type, _Mapped_args_t...>
+              && (same_as<remove_cvref_t<_K &&>, key_type>
+                  || (constructible_from<key_type, _K&&> && _Compare_is_transparent<key_compare>) )
+    {
+        const const_iterator _Begin = cbegin();
+        const const_iterator _End   = cend();
+
+        // Adjust _Position
+        if (_Position == _End || !_Key_compare(*(_Position._Key_it), _Key)) {
+            // _Val should be inserted at or before _Position
+            if (_Position == _Begin || !_Key_compare(_Key, *((_Position - 1)._Key_it))) {
+                // _Val should be inserted at or after _Position
+                // The hint _Position is accurate
+            } else {
+                auto _Key_Position = _STD upper_bound(_Data.keys.cbegin(), _Position._Key_it, _Key, _Key_compare);

```

Comment:
For `flat_map` (or `flat_set`), to decide the existence of key we should always firstly find `lower_bound(key)`. `upper_bound` can result in broken state.

Reply:
Arrr you are right. I plagiarized it from `flat_set` when I wrote this part.

Reply:
I rewrote this function. Hope it's correct now.